OsmAnd-java/src/main/java/net/osmand/router/RoutePlannerFrontEnd.java

```diff
@@ -1078,53 +1067,75 @@ private void makeSegmentPointPrecise(RouteSegmentResult routeSegmentResult, LatL
-			return false;
-		} else {
-			log.info("Route segment found " + f.road);
-			res.add(f);
-			return true;
+	private RouteCalcResult findSegmentPnt(final RoutingContext ctx, LatLon pnt, int i, List<RouteSegmentPoint> st)
+			throws IOException {
+		while (i >= st.size()) {
+			st.add(null);
+		}
+		if (st.get(i) == null) {
+			RouteSegmentPoint s = findRouteSegment(pnt.getLatitude(), pnt.getLongitude(), ctx, null,
+					ctx.startTransportStop);
+			if (s == null) {
+				ctx.calculationProgress.segmentNotFound = i;
+				return new RouteCalcResult((i == 0 ? "Start" : "Target") + " point is not located: " + pnt);
+			}
+			st.set(i, s);
 		}
+		return null;
+	}
 
+	private RouteCalcResult searchRouteAndPrepareTurns(final RoutingContext ctx, RouteSegmentPoint s,
+			RouteSegmentPoint e, PrecalculatedRouteDirection routeDirection) throws IOException, InterruptedException {
+		List<RouteSegmentPoint> ls = new ArrayList<>();
+		ls.add(s);
+		ls.add(e);
+		return searchRouteAndPrepareTurns(ctx, s.getPreciseLatLon(), e.getPreciseLatLon(), ls, 0, routeDirection);
 	}
 
-	private RouteCalcResult searchRouteInternalPrepare(final RoutingContext ctx, RouteSegmentPoint start, RouteSegmentPoint end,
-	                                                            PrecalculatedRouteDirection routeDirection) throws IOException, InterruptedException {
+	private RouteCalcResult searchRouteAndPrepareTurns(final RoutingContext ctx, LatLon start, LatLon end,
+			List<RouteSegmentPoint> points, int i, PrecalculatedRouteDirection routeDirection)
+			throws IOException, InterruptedException {
 		RouteSegmentPoint recalculationEnd = getRecalculationEnd(ctx);
+		RouteSegmentPoint s = null; // null for native
+		RouteSegmentPoint e = null; // null for native
 		if (recalculationEnd != null) {
-			ctx.initStartAndTargetPoints(start, recalculationEnd);
+			e = recalculationEnd;
+		}
+		if (ctx.nativeLib != null) {
+			ctx.initStartEndPoints(s != null ? s.getPreciseLatLon() : start, e != null ? e.getPreciseLatLon() : end, null);
 		} else {
-			ctx.initStartAndTargetPoints(start, end);
+			RouteCalcResult err = findSegmentPnt(ctx, start, i, points);
+			if (err != null) {
+				return err;
+			} else {
+				s = points.get(i);
+			}
+			err = findSegmentPnt(ctx, end, i + 1, points);
+			if (err != null) {
+				return err;
+			} else {
+				e = points.get(i + 1);
+			}
+			ctx.initStartAndTargetPoints(s, e);
 		}
 		if (routeDirection != null) {
 			ctx.precalculatedRouteDirection = routeDirection.adopt(ctx);
 		}
+		List<RouteSegmentResult> result;
 		if (ctx.nativeLib != null) {
-			ctx.startX = start.preciseX;
-			ctx.startY = start.preciseY;
-			ctx.startRoadId = start.road.id;
-			ctx.startSegmentInd  = start.segStart;
-			ctx.targetX = end.preciseX;
-			ctx.targetY = end.preciseY;
-			ctx.targetRoadId = end.road.id;
-			ctx.targetSegmentInd  = end.segStart;
-			return runNativeRouting(ctx, recalculationEnd, null);
+			RouteSegmentResult[] res = runNativeRouting(ctx, null);
```

Comment:
Lost ctx.targetRoadId etc. for native routing

Reply:
> Lost ctx.targetRoadId etc. for native routing
Target road id in Java couldn't be set because addSegment should happen in C++, to speed up routing. It was working before as well.

Reply:
Actually, not set in C++. The start/end RoadId & start/end SegmentInd were set from Java side because java_wrap calls BRP directly w/o C++ RoutePlannerFrontend functions (where is the setters located). So we need this vars passed directly from Java.

Reply:
fixed